File: togo/internal/cli/annotations.go
package cli

// Port of src/cli/annotations.ts. GitHub Actions / Forgejo / Gitea
// workflow-command annotations: when cappu runs inside a runner that parses
// GitHub-style workflow commands, errors and warnings are echoed as
// `::error file=...,line=...::message` so they surface as inline annotations,
// in addition to the normal stderr output.
// https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

// AnnotationLocation is the optional file/line/col of an annotation. A zero
// field is treated as absent (line/col 0 is never a real 1-based position).
type AnnotationLocation struct {
	File   string
	Line   int
	Column int
}

// AnnotationsEnabled reports whether the current runner understands
// GitHub-style workflow commands. Forgejo/Gitea Actions also set
// GITHUB_ACTIONS=true for compatibility, but we check their own vars too so the
// intent is explicit. Bare CI=true is not a trigger: there is no generic-CI
// annotation format, and emitting GitHub syntax in a non-GitHub runner would
// just be noise. The env lookup is a parameter so it stays testable.
func AnnotationsEnabled(env func(string) string) bool {
	return env("GITHUB_ACTIONS") == "true" ||
		env("FORGEJO_ACTIONS") == "true" ||
		env("GITEA_ACTIONS") == "true"
}

// https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands#example-setting-an-error-message
var dataEscaper = strings.NewReplacer("%", "%25", "\r", "%0D", "\n", "%0A")
var propEscaper = strings.NewReplacer("%", "%25", "\r", "%0D", "\n", "%0A", ":", "%3A", ",", "%2C")

// FormatAnnotation builds one workflow-command line (no trailing newline).
func FormatAnnotation(severity, message string, loc AnnotationLocation) string {
	var props []string
	if loc.File != "" {
		props = append(props, "file="+propEscaper.Replace(loc.File))
	}
	if loc.Line != 0 {
		props = append(props, "line="+propEscaper.Replace(strconv.Itoa(loc.Line)))
	}
	if loc.Column != 0 {
		props = append(props, "col="+propEscaper.Replace(strconv.Itoa(loc.Column)))
	}
	head := "::" + severity
	if len(props) > 0 {
		head += " " + strings.Join(props, ",")
	}
	return head + "::" + dataEscaper.Replace(message)
}

// emitAnnotation echoes an annotation to stderr when running under a supporting
// CI runner.
func emitAnnotation(severity, message string, loc AnnotationLocation) {
	if AnnotationsEnabled(os.Getenv) {
		fmt.Fprintln(os.Stderr, FormatAnnotation(severity, message, loc))
	}
}

// EmitErrorAnnotation is the exported, location-less error variant for callers
// outside package cli (the config-load error path in package main).
func EmitErrorAnnotation(message string) {
	emitAnnotation("error", message, AnnotationLocation{})
}

File: src/cli/annotations.ts
// GitHub Actions / Forgejo / Gitea workflow-command annotations. When cappu runs
// inside a runner that parses GitHub-style workflow commands, errors and warnings
// are echoed as `::error file=...,line=...::message` so they surface as inline
// annotations, in addition to the normal stderr output.
// https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands

export type AnnotationSeverity = "error" | "warning";

export interface AnnotationLocation {
  file?: string;
  line?: number;
  column?: number;
}

/**
 * Whether the current runner understands GitHub-style workflow commands.
 * Forgejo/Gitea Actions also set GITHUB_ACTIONS=true for compatibility, but we
 * check their own vars too so the intent is explicit. Bare CI=true is not a
 * trigger: there is no generic-CI annotation format, and emitting GitHub syntax
 * in a non-GitHub runner would just be noise.
 */
export function annotationsEnabled(env: NodeJS.ProcessEnv = process.env): boolean {
  return (
    env.GITHUB_ACTIONS === "true" || env.FORGEJO_ACTIONS === "true" || env.GITEA_ACTIONS === "true"
  );
}

// https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands#example-setting-an-error-message
function escapeData(s: string): string {
  return s.replaceAll("%", "%25").replaceAll("\r", "%0D").replaceAll("\n", "%0A");
}

function escapeProp(s: string): string {
  return escapeData(s).replaceAll(":", "%3A").replaceAll(",", "%2C");
}

/** Format one workflow-command annotation line (no trailing newline). */
export function formatAnnotation(
  severity: AnnotationSeverity,
  message: string,
  loc?: AnnotationLocation,
): string {
  const props: string[] = [];
  if (loc?.file !== undefined) props.push(`file=${escapeProp(loc.file)}`);
  if (loc?.line !== undefined) props.push(`line=${escapeProp(String(loc.line))}`);
  if (loc?.column !== undefined) props.push(`col=${escapeProp(String(loc.column))}`);
  const head = props.length > 0 ? `::${severity} ${props.join(",")}` : `::${severity}`;
  return `${head}::${escapeData(message)}`;
}

/** Echo an annotation to stderr when running under a supporting CI runner. */
export function emitAnnotation(
  severity: AnnotationSeverity,
  message: string,
  loc?: AnnotationLocation,
): void {
  if (annotationsEnabled()) {
    process.stderr.write(`${formatAnnotation(severity, message, loc)}\n`);
  }
}
